File: apps/ios/ProviiIssuerDemo/ProviiIssuerDemo/DemoTokenManager.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2025 Provii

// Demo Token Manager for Provii Demo Apps
//
// Fetches and caches the rotating demo token from playground.provii.app.
// This token authenticates requests to the demo backends and prevents
// unauthorised bot/spam access.
//
// Token format: demo_token_v1_<YYYYMMDD>_<16-char-hmac>
// Tokens are valid for 48 hours (today + yesterday) for timezone handling.

import Foundation

/// Manages fetching and caching of the rotating demo authentication token.
///
/// The token is refreshed automatically when it approaches expiry. Cached tokens
/// are reused until they fall within one hour of expiration.
class DemoTokenManager {
    /// Shared singleton instance.
    static let shared = DemoTokenManager()

    private let tokenEndpoint = "https://playground.provii.app/v1/config/demo-token"
    private var cachedToken: String?
    private var tokenExpiresAt: TimeInterval = 0

    private init() {}

    /// Fetches the current demo token, returning a cached value if still valid.
    ///
    /// - Returns: The demo token string for use in X-Demo-Token headers
    /// - Throws: ``DemoTokenError`` if the token endpoint is unreachable or returns an error
    func getToken() async throws -> String {
        let now = Date().timeIntervalSince1970

        // Return cached token if still valid (with 1 hour buffer)
        if let token = cachedToken, tokenExpiresAt > now + 3600 {
            return token
        }

        // SECURITY: Token fetched over HTTPS from the demo configuration endpoint
        guard let url = URL(string: tokenEndpoint) else {
            throw DemoTokenError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw DemoTokenError.fetchFailed
        }

        let tokenResponse = try JSONDecoder().decode(DemoTokenResponse.self, from: data)

        cachedToken = tokenResponse.token
        tokenExpiresAt = Double(tokenResponse.expiresAt)

        return tokenResponse.token
    }
}

// MARK: - Supporting Types

/// Errors that can occur when fetching the demo token.
enum DemoTokenError: Error, LocalizedError {
    /// The token endpoint URL is malformed.
    case invalidURL
    /// The token fetch request failed or returned a non-2xx status.
    case fetchFailed

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid demo token endpoint URL"
        case .fetchFailed:
            return "Failed to fetch demo token"
        }
    }
}

/// Response structure from the demo token endpoint.
private struct DemoTokenResponse: Codable {
    let token: String
    let expiresAt: Int
    let cacheSeconds: Int

    enum CodingKeys: String, CodingKey {
        case token
        case expiresAt = "expires_at"
        case cacheSeconds = "cache_seconds"
    }
}

File: apps/android/issuer/app/src/main/kotlin/app/provii/demo/issuer/api/DemoTokenManager.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2025 Provii

/**
 * Demo Token Manager for Provii Demo Apps.
 *
 * Fetches and caches the rotating demo token from playground.provii.app.
 * This token is used to authenticate requests to the demo backends
 * and prevent unauthorised bot/spam access.
 *
 * Token format: demo_token_v1_<YYYYMMDD>_<16-char-hmac>
 * Tokens are valid for 48 hours (today + yesterday) for timezone handling.
 */

package app.provii.demo.issuer.api

import android.util.Log
import app.provii.demo.issuer.BuildConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.android.Android
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.accept
import io.ktor.client.request.get
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private const val TAG = "DemoTokenManager"
private const val TOKEN_ENDPOINT = "https://playground.provii.app/v1/config/demo-token"
private const val MILLIS_PER_SECOND = 1000
private const val TOKEN_REFRESH_BUFFER_SECONDS = 3600

@Serializable
private data class DemoTokenResponse(
    val token: String,
    @SerialName("expires_at") val expiresAt: Long,
    @SerialName("cache_seconds") val cacheSeconds: Int
)

/**
 * Singleton manager for demo authentication tokens.
 *
 * Usage:
 * ```kotlin
 * val token = DemoTokenManager.getToken()
 * // Add to your request headers: "X-Demo-Token" to token
 * ```
 */
object DemoTokenManager {
    private var cachedToken: String? = null
    private var tokenExpiresAt: Long = 0

    private val client = HttpClient(Android) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
            })
        }
    }

    /**
     * Get the current demo token, fetching a new one if needed.
     *
     * // SECURITY: Token is fetched over HTTPS and cached with a 1-hour buffer
     * // before expiry to avoid using stale credentials.
     *
     * @return The demo token string
     * @throws java.io.IOException if network request fails
     * @throws IllegalStateException if the server returns a non-200 response
     */
    suspend fun getToken(): String {
        val now = System.currentTimeMillis() / MILLIS_PER_SECOND

        // Return cached token if still valid (with 1 hour buffer)
        cachedToken?.let { token ->
            if (tokenExpiresAt > now + TOKEN_REFRESH_BUFFER_SECONDS) {
                return token
            }
        }

        // Fetch new token
        return try {
            val response = client.get(TOKEN_ENDPOINT) {
                accept(ContentType.Application.Json)
            }

            check(response.status == HttpStatusCode.OK) {
                "Failed to fetch demo token: ${response.status}"
            }

            val tokenResponse = response.body<DemoTokenResponse>()
            cachedToken = tokenResponse.token
            tokenExpiresAt = tokenResponse.expiresAt

            tokenResponse.token
        } catch (e: java.io.IOException) {
            if (BuildConfig.DEBUG) {
                Log.e(TAG, "Network error fetching demo token", e)
            }
            throw e
        }
    }
}
